allennlp/data/semparse/worlds/world.py

```diff
@@ -206,6 +209,75 @@ def get_action_sequence(self, expression: Expression) -> List[str]:
         return self._get_transitions(expression,
                                      ["%s -> %s" % (types.START_TYPE, expression.type)])
 
+    def get_logical_form(self,
+                         action_sequence: List[str],
+                         add_var_function: bool = True) -> str:
+        """
+        Takes an action sequence and constructs a logical form from it. This is useful if you want
+        to get a logical form from a decoded sequence of actions generated by a transition based
+        semantic parser.
+
+        Parameters
+        ----------
+        action_sequence : ``List[str]``
+            The sequence of actions as strings (eg.: ``['@START@ -> t', 't -> <e,t>', ...]``).
+        add_var_function : ``bool`` (optional)
+             ``var`` is a special function that some languages use within lambda functions to
+             indicate the use of a variable (eg.: ``(lambda x (fb:row.row.year (var x)))``). Due to
+             the way constrained decoding is currently implemented, it is easier for the decoder to
+             not produce these functions. In that case, setting this flag adds the function in the
+             logical form even though it is not present in the action sequence.
+        """
+        # TODO(mattg,pradeep): The way "reverse" is handled here is a mess.  To make this more
+        # general, we need to fix some things in the type system.
+        split_actions = [action.split(" -> ") for action in action_sequence]
+        terminals: List[Tuple[str, int]] = []  # terminal name and the number of arguments it takes
+        for left_side, right_side in split_actions:
+            right_side_is_number = False
+            try:
+                float(right_side)
+                right_side_is_number = True
+            except ValueError:
+                pass
+            if '[' in right_side:
+                if 'lambda ' in right_side:
+                    # "lambda x/y/...". Removing single quotes surrounding the string.
+                    terminals.append((right_side[1:-1].split(', ')[0].replace("'", ""),
+                                      self._infer_num_arguments(left_side)))
+            elif len(right_side) > 1 or right_side in ['x', 'y', 'z', '<', '-'] or right_side_is_number:
+                if right_side == "var" and add_var_function:
+                    raise RuntimeError("You wanted to add var, but you didn't need to!")
+                if right_side in ['x', 'y', 'z'] and add_var_function:
+                    right_side = f"(var {right_side})"
+                terminals.append((right_side, self._infer_num_arguments(left_side)))
+        partial_logical_forms: List[str] = []
+        terminals = list(reversed(terminals))
+        for i, (terminal, num_args) in enumerate(terminals):
```

Comment:
Can you add a comment here describing the special way in which you're handling `reverse` and why?